sqlcapture/capture.go

```diff
@@ -454,50 +417,70 @@ func (c *Capture) streamToWatermark(ctx context.Context, replStream ReplicationS
 }
 
 func (c *Capture) streamForever(ctx context.Context, replStream ReplicationStream) error {
-	return c.streamToWatermarkWithOptions(ctx, replStream, nonexistentWatermark, true)
+	logrus.Info("streaming replication events indefinitely")
+	for ctx.Err() == nil {
+		// Spawn a worker goroutine which will write the watermark value after the appropriate interval.
+		var watermark = uuid.New().String()
+		var group, workerCtx = errgroup.WithContext(ctx)
+		group.Go(func() error {
+			select {
+			case <-workerCtx.Done():
+				return workerCtx.Err()
+			case <-time.After(heartbeatWatermarkInterval):
+				if err := c.Database.WriteWatermark(workerCtx, watermark); err != nil {
+					return fmt.Errorf("error writing next watermark: %w", err)
+				}
+			}
+			return nil
+		})
+
+		// Perform replication streaming until the watermark is reached (or the
```

Comment:
you thought about running this with `group.Go` ?

In this context, because your using `workerCtx` I agree this implementation seems correct. It just threw my pattern-matching instinct which was looking for multiple calls to `group.Go` with a single blocking Wait that's looking for the first returned error.

Though I suppose a difference is that an error to write the watermark could be buried by a secondary context cancellation error of `streamCatchup` ?

Reply:
That's a good point, I hadn't thought about the potential for `streamCatchup` to return a useless context-cancellation error and hide the real failure. I'll change this so that both operations run via `group.Go`.